### DISCUSSION TITLE
[FEATURE] opt-in install and postinstall scripts

### DISCUSSION THREAD

**Discussion by koba04** [npm/rfcs#80] [General]
# What / Why

`install` and `postinstall` are npm scripts that run while `npm install ${package}` or `npm ci`.
It means that we allow the package author to run any scripts on our machine, which is definitely dangerous if we don't recognize what packages are installed as dependencies.
Unfortunately, it might be impossible to recognize all the dependencies on your project so we run scripts we don't know while `npm install`.

To avoid this, we can use the `--ignore-scripts` option, which disables all npm scripts while installing dependencies.
But sometimes the scripts are required to work the packages.

I've created an npm package to print all these scripts in your project.
https://github.com/koba04/install-scripts

But I think it would be nice if npm has the mechanism to opt-in `install` and `postinstall` scripts.
I guess there are some ways to do this. The followings are ideas on me. Please let me know If you have any ideas :pray:

### `allowedScripts` in `package.json`

This field is to allow packages to run npm scripts.
With this option, we can share the information that we allow packages to run scripts through the package.json.
It might be better this option is enabled only with `--allowed-scripts` option.

```json
{
  "allowedScripts": {
    "fsevents": "^1.0.0",
    "puppeteer": "*"
  }
}
```

### `--confirm-scripts`

This option is to show a prompt whether you run the script or not.
This is useful if you know what the script would do.

```sh
Do you run the install script for fsevents?
y/n
```

### other option

This is another feature request for `--ignore-scripts`.
I hope that npm prints all scripts not to run while installing with `--ignore-scripts` like https://github.com/koba04/install-scripts.

## When

* `npm install` or `npm ci`

## Where

* npm public registry

### Current Behavior

* run all `install` and `postinstall` scripts or don't run any scripts with `--ignore-scripts` option.

### Expected Behavior

* We can opt-in scripts to be able to run.

## Who

* n/a

## References

* https://github.com/yarnpkg/yarn/issues/7761

**Comment by koba04** [comment]
Yarn v2 has a similar feature using `dependenciesMeta`.
https://dev.to/arcanis/introducing-yarn-2-4eh1#perpackage-build-configuration

**Reply by DanielRuf** [reply]
We should go from a "allow everything by default" to a "deny everything by default" to protect most developers against such attacks with `pkg.scripts` by default. See also my comment regarding hashes (strict mode, that's something that I would like to use).

Because so far the registry is way too big, no releases are actively checked (I did some experiments in the past and reuploaded disabled malware that was was accidentally found by Sonatype and removed by npm after 8 months, see https://blog.sonatype.com/nexus-intelligence-insights-sonatype-2018-0413) by npm / GitHub, no access for threat hunters to run YARA rules on the package corpus, ... =(

**Comment by DanielRuf** [comment]
> ```json
> {
>   "whitelistScripts": [
>     "fsevents",
>     "puppeteer"
>   ]
> }
> ```

I would like to see some solution like hashes for scripts (only hash `pkg.scripts` as text). Otherwise these could be weaponized in the future and the scripts could be changed to deliver malware. See how Comodo, Zonealarm and others did this (strict whitelist, based on hashes / data).

Something similar like the SRI hashes in browsers. People also wanted this feature in Deno but it looks like it was never implemented: https://github.com/denoland/deno/issues/200

**Comment by DanielRuf** [comment]
Another recommendation: please enforce 2FA for all my packages when I setup 2FA for my npm account. I have to do this manually for every single package.

**Comment by sandstrom** [comment]
I'm guessing this would help against npm packages executed only in the browser. 

But for packages running in node, they still have disk access + can execute on the shell, no? So in that later case shutting down scripts wouldn't help much?

**Reply by DanielRuf** [reply]
> I'm guessing this would help against npm packages executed only in the browser.

No, the `--ignore-scripts` affects the `npm install` command. The "SRI" part from my side was just an example for using hashes in general, nothing more.

**Reply by DanielRuf** [reply]
> But for packages running in node, they still have disk access + can execute on the shell, no?

There are also policies available in Node / npm: https://nodejs.org/api/policy.html
But this is about a different RFC.

**Reply by DanielRuf** [reply]
> So in that later case shutting down scripts wouldn't help much?

It does, because with `--ignore-scripts` no `postinstall` script is executed at all.

**Reply by sandstrom** [reply]
My point is that any npm package running under node will usually have access to entire developer machine anyway.

**Reply by DanielRuf** [reply]
> My point is that any npm package running under node will usually have access to entire developer machine anyway.

Sure, but that is another problem which will be covered by policies in the future. This RFC here is only about the default setting for running scripts in pkg.scripts (package.json) like postinstall (so we don't have to use `--ignore-scripts` for every command.

**Reply by mohe2015** [reply]
I think this change here should still be done but I think this is a larger issue. The dependency chains need to be reduced, only actively maintained packages should be in them for popular packages and packages should be built from Github/the original source code.

**Reply by DanielRuf** [reply]
@mohe2015 as others pointed out in the ua-parser-js issue, you can not enforce that

> packages should be built from Github/the original source code

`npm pack` and `npm publish` were not designed like this

**Reply by cmdcolin** [reply]
>My point is that any npm package running under node 

note that a high proportion of npm packages will never even run under node...designed for web

**Reply by ljharb** [reply]
@cmdcolin i'm pretty sure that in fact is a very, very low proportion. The vast majority of JS packages on npm works universally, in my experience.

**Reply by cmdcolin** [reply]
sure. I guess I am likely wrong in that sense. I guess I mean to say that *I, as a web dev* would not be running many of the packages I am installing under node

**Reply by koba04** [reply]
> But for packages running in node, they still have disk access + can execute on the shell, no? So in that later case shutting down scripts wouldn't help much?

Yes, they can execute any scripts when they are used as `require("some-package")`, so security issues are still remained. But this is often used as an attack vector and install scripts are more dangerous than the case because this allows all installed packages to run any scripts without using them. In this case, typing the wrong package name also causes serious issues because it gives permissions to run arbitrary scripts to un-intentional packages.

**Reply by pmoleri** [reply]
@sandstrom your concerns are valid, but this RFC is about npm running non-sandboxed scripts on package installation, while what you describe is an node limitation, that as [DanielRuf mentioned ](https://github.com/npm/rfcs/discussions/80#discussioncomment-1531399), will be covered by node.
What I want to say is that I wouldn't want to see this great proposal delayed by this argument.

**Comment by koba04** [comment]
https://github.com/npm/rfcs/pull/488 is another proposal related to this. It changes the default behavior of install scripts.

Note that I've updated the `whitelistScripts` field from an array to an object to be able to specify versions.

**Reply by DanielRuf** [reply]
> to an object to be able to specify versions

Very good idea. As packages on the registry can not be changed after they are uploaded / published, this sounds like a good way to pin the allowed scripts to specific versions. But the content of the scripts and new scripts still have to be checked by the users then, but this is another issue.

**Comment by felipeplets** [comment]
For semantic reasons I would recommend the usage of `allowlist` over `whitelist`

Thanks for the RFC @koba04 
@DanielRuf I second your thoughts on going from allow to deny by default.

**Reply by koba04** [reply]
Thank you! I've updated the usage in RFC.

**Reply by mohe2015** [reply]
Maybe just `allowedScripts`?

**Comment by bregenspan** [comment]
Just to note some prior art, PNPM implemented this here: https://github.com/pnpm/pnpm/pull/8897

**Comment by cmawhorter** [comment]
Personally, I'd like to see more verbose info during confirmation i.e. printing out the script that will be executed, so I can inspect it.

Also, allowedScripts supporting non-specific versions seems to create a false sense of security IMO. The most recent problems I'm aware of with postinstall are related to an existing pkg adding malware, which this doesn't really address.

From a quick read of comments here and a few related npm issues, the opposition seems to be that deps already have local access and can execute arbitrary code, so why bother? That's sorta true, but in my mind, the more likely scenario here is that some deep dep in some trusted pkg adds something bad. That bad pkg isn't executed in app code anywhere, but a postinstall script still would be.

**Reply by tafelnl** [reply]
> the opposition seems to be that deps already have local access and can execute arbitrary code, so why bother?

There's actually a couple of very good arguments that counter this statement. tolmasky outlined some in https://github.com/npm/rfcs/pull/488#issuecomment-964714944

**Comment by manzoorwanijk** [comment]
The recent ["Shai-Hulud" supply chain attack on the npm ecosystem](https://socket.dev/blog/ongoing-supply-chain-attack-targets-crowdstrike-npm-packages) has been successful in part due to postinstall scripts that have allowed it to rapidly self-propagate.

That makes this feature even more important.

**Comment by jakubsemerak** [comment]
Another one attack using postinstall scripts dropped yesterday https://www.stepsecurity.io/blog/axios-compromised-on-npm-malicious-versions-drop-remote-access-trojan

**Comment by everett1992** [comment]
RFC [Make npm install scripts opt-in](https://github.com/npm/rfcs/pull/488) seems to be blocked because it is a breaking change. I created a new RFC [Add option to require install script approval](https://github.com/npm/rfcs/pull/861) proposing an opt-in allowlist.

**Comment by Daniel15** [comment]
It's coming: https://github.com/orgs/community/discussions/198547
